### PR TITLE
[2/n] add `Stateful` functionality support for Callbacks

### LINKED ISSUE #11429: [RFC] Standardize all stateful components on `state_dict`/`load_state_dict`
## Proposed refactor
Standardize all stateful components on `state_dict`/`load_state_dict`.

### Background

PyTorch convention uses `state_dict`/`load_state_dict` for gathering and loading object state. In lightning, some components follow this convention, while other components do not (see Appedix: current state gathering section)

### Motivation
 
Each component should contribute saving/loading their own state with the same APIs.

1. independently contributing/loading one’s own local component state (aligning with PyTorch primitives: `state_dict`/`load_state_dict`)
2. operating and depending on global component state (Lightning `CheckpointHooks`: `on_save/load_checkpoint`)

**This issue is focused on aligning all components on 1**. Following this convention will allow consistency across Lightning components and consistency with PyTorch conventions. Any stateful component can simply implement their own `state_dict`/`load_state_dict` methods to contribute their own state.
For now **2** is only adjusted as needed (Callbacks).

### Pitch

Lightning already has this Stateful Protocol in `auto_restart.py`. We can move this `_SupportsStateDict` out to the more central `core/hooks.py` file:

https://github.com/PyTorchLightning/pytorch-lightning/blob/34c62da37dc1ed9a1de7e023c690c7528ee56c60/pytorch_lightning/utilities/auto_restart.py#L638-L646

```py
@runtime_checkable
class Stateful(Protocol):
    """This class is used to detect if an object is stateful using `isinstance(obj, Stateful)`."""

    def state_dict(self) -> Dict[str, Any]:
        ...

    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
        ...
```
### Additional context

Part of #7740

### Appendix: current state gathering:
Specifically, current components contribute state in the following different ways:
1. Some components contribute saving/loading their own state with only `on_save/load_checkpoint` [DataModule, PrecisionPlugin, Callbacks]
a. [DataModule, PrecisionPlugin] use `on_save/load_checkpoint` from `CheckpointHooks` https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/core/hooks.py#L765-L807
b. [Callbacks] contribute saving/loading their own state with different `on_save/load_checkpoint` hook signatures and functionality https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/callbacks/base.py#L293-L323
c. though [Loops] falls under 3. below, noting here that Loops also have their own `on_save/load_checkpoint` methods with different signatures. https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/loops/base.py#L253-L262
2. Some components contribute saving/loading their own state with only `state_dict`/`load_state_dict` calls [Optimizers, LR schedulers]
3. Some components have both [LightningModule, Loops]
#### Save aligning on state_dict
`dump_checkpoint`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L310-L393
becomes:
```py
def dump_checkpoint(self, weights_only: bool = False) -> dict:
   ...
   # dump callbacks
   # checkpoint["callbacks"] = self.trainer._call_callbacks_on_save_checkpoint(checkpoint)
   # becomes
   checkpoint["callbacks"] = self.trainer._call_callbacks_state_dict()
   
   ...
   # precision plugin
   # self.trainer.precision_plugin.on_save_checkpoint(checkpoint)
   # becomes
   prec_plugin = self.trainer.precision_plugin
   checkpoint[prec_plugin.__class__.__name__] = self.trainer.precision_plugin.state_dict()

   ...
   # give the model a chance to dump a few things
   # model.on_save_checkpoint(checkpoint)
   # if self.trainer.datamodule is not None:
   #    self.trainer.datamodule.on_save_checkpoint(checkpoint)
   # becomes
   
   # datamodule state
   dm = self.trainer.datamodule
   if dm is not None:
       checkpoint[dm.__class__.__name__] = dm.state_dict()
       
   # on_save_checkpoint calls
   model.on_save_checkpoint(checkpoint)
   dm.on_save_checkpoint(checkpoint)
   prec_plugin.on_save_checkpoint(checkpoint)
   for callback in self.trainer.callbacks:
       callback.on_save_checkpoint(self.trainer, model, checkpoint)
```
#### Load aligning on load_state_dict
see component Load sections below

#### Base Class

https://github.com/PyTorchLightning/pytorch-lightning/blob/59a7ba760548baadf6dbb30864b54cb01c7225a3/pytorch_lightning/callbacks/base.py#L293-L323

becomes
BC: `on_save_checkpoint` returns `None` instead of `dict`
BC: `on_load_checkpoint` arg takes entire `checkpoint` dict instead of `callback_state`
```py
def state_dict(self) -> Dict[str, Any]:
    return {}

def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
    pass

def on_save_checkpoint(
        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
    ) -> None:
    """Called by Lightning when saving a checkpoint to give you a chance to store or customize anything
    else you might want to save.
    Args:
        trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
        pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
        checkpoint: the checkpoint dictionary that will be saved.
    """
    pass

def on_load_checkpoint(
    self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
) -> None:
    """Called by Lightning when loading a checkpoint to give you a chance to reload or customize anything
    else you may have saved in on_save_checkpoint.
    Args:
        trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
        pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
        checkpoint: entire loaded checkpoint dictionary
    """
    pass
```
#### Save
https://github.com/PyTorchLightning/pytorch-lightning/blob/85304d4672a9ed24a16f7f5b2abaa34148ab86f4/pytorch_lightning/trainer/trainer.py#L1603
```py
def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[str, dict]:
    ...
    state = callback.on_save_checkpoint(self, self.lightning_module, checkpoint)
```
becomes
```py
def _call_callbacks_state_dict(self) -> Dict[str, dict]:
   ...
   state = callback.state_dict()
```
#### Load
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/trainer.py#L1652
```py
def _call_callbacks_on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
    ...
    callback.on_load_checkpoint(self, self.lightning_module, state)
```
becomes
```py
def _call_callbacks_load_state_dict(self, checkpoint: Dict[str, Any]) -> None:
    ...
    callback.load_state_dict(state)
```
`restore_callbacks`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L205
becomes
```py
for callback in self.trainer.callbacks:
    callback.on_load_checkpoint(self.trainer, self.trainer.lightning_module, self._loaded_checkpoint)
self.trainer._call_callbacks_load_state_dict(self._loaded_checkpoint)
```
#### Callback classes
update `timer`, `pruning`, `model_checkpoint`, `finetuning`, `early_stopping` to use `state_dict`/`load_state_dict` instead of `on_save/load_checkpoint`.

#### Base Class
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/precision/precision_plugin.py
add dummy `state_dict/load_state_dict`:
```py
def state_dict(self) -> Dict[str, Any]:
    return {}
def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
    pass
```
#### Load
`restore_training_state`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L189-L190
becomes
```py
# restore precision plugin (scaler etc.)
prec_plugin = self.trainer.precision_plugin
prec_plugin.on_load_checkpoint(self._loaded_checkpoint)
if prec_plugin.__class__.__name__ in self._loaded_checkpoint:
    prec_plugin.load_state_dict(self._loaded_checkpoint[prec_plugin.__class__.__name__])
```
#### Precision Plugin classes

Update `apex_amp`, `native_amp` to use `state_dict/load_state_dict` instead of `on_save/load_checkpoint`

#### Base Class
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/datamodule.py
add dummy `state_dict/load_state_dict`:
```py
def state_dict(self) -> Dict[str, Any]:
    return {}
def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
    pass
```
#### Load

`restore_datamodule`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L151
becomes
```py
datamodule.on_load_checkpoint(self._loaded_checkpoint)
if datamodule.__class__.__name__ in self._loaded_checkpoint:
    datamodule.load_state_dict(self._loaded_checkpoint[datamodule.__class__.__name__])
```

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @rohitgr7 @ananthsub @ninginthecloud

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Part of / Fixes #11429 : introduce Stateful Callbacks
part 2. of the [full implementation PR reference](https://github.com/PyTorchLightning/pytorch-lightning/pull/11887#issuecomment-1045650933)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

File: tests/callbacks/test_callbacks.py [hunk 1]
```diff
@@ -137,6 +138,41 @@ def test_resume_callback_state_saved_by_type(tmpdir):
...
+
+    def on_load_checkpoint(self, trainer, pl_module, callback_state):
+        self.state = callback_state["state"]
+
+
+def test_resume_callback_state_saved_by_type_hooks(tmpdir):
+    """Test that a legacy checkpoint that didn't use a state key before can still be loaded, using
+    on_save/load_checkpoint."""
+    # TODO: remove on_save_checkpoint() -> dict
+    # and on_load_checkpoint(callback_state) support in v1.8
```
Comment (awaelchli) [inline tests/callbacks/test_callbacks.py:160]: Once again I'm struggling to understand why we have a todo to remove the hook. 
I thought we proposed to standardize the `on_*_checkpoint` methods to take the full checkpoint dict as an argument and remove the return value.

The todo should say that the test gets removed?

Comment (jjenniferdai) [reply tests/callbacks/test_callbacks.py:160]: todo is only to remove on_save_checkpoint **-> dict** and on_load_checkpoint(**callback_state**) support.
will add a line clarifying "in favor of on_save_checkpoint **-> None** and on_load_checkpoint(**checkpoint**)"

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -319,7 +320,8 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
                 'epoch':                     training epoch
                 'global_step':               training global step
                 'pytorch-lightning_version': The version of PyTorch Lightning that produced this checkpoint
-                'callbacks':                 "callback specific state"[] # if not weights_only
+                'callbacks':                                  # if not weights_only
+                    {callback.state_key: callback.on_save_checkpoint or callback.state_dict}
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:324]:
```suggestion
```
this comment does not have enough context for someone new reading this to understand what this means. i would recommend no adding this here, but rather elsewhere in the checkpoint connector closer to where this switch happens

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1661,19 +1661,28 @@ def _on_train_batch_end(self, outputs: STEP_OUTPUT, batch, batch_idx, dataloader
...
-    def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[str, dict]:
-        """Called when saving a model checkpoint.
+    def _call_callbacks_state_dict(self) -> Dict[str, dict]:
+        """Called when saving a model checkpoint, calls and returns every callback's `state_dict`, keyed by
+        `Callback.state_key`."""
+        callback_state_dicts = {}
+        for callback in self.callbacks:
+            state_dict = callback.state_dict()
+            if state_dict:
+                callback_state_dicts[callback.state_key] = state_dict
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1671]: while you're here, please profile the callback state_dict calls so we don't have to add this later

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1661,19 +1661,28 @@ def _on_train_batch_end(self, outputs: STEP_OUTPUT, batch, batch_idx, dataloader
...
         """
-        callback_states = {}
         for callback in self.callbacks:
             # TODO: Add profiling for on_save_checkpoint hook
             state = callback.on_save_checkpoint(self, self.lightning_module, checkpoint)
             if state:
-                callback_states[callback.state_key] = state
-        return callback_states
+                # TODO: Add deprecation warning if state is returned (see reference PR #11887)
+                checkpoint["callbacks"][callback.state_key] = state
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1685]: same here

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1703,6 +1712,19 @@ def _call_callbacks_on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None
...
+        callback_states: Dict[Union[Type, str], Dict] = checkpoint.get("callbacks")
+
+        if callback_states is None:
+            return
+
+        for callback in self.callbacks:
+            state = callback_states.get(callback.state_key, callback_states.get(callback._legacy_state_key))
+            if state:
+                state = deepcopy(state)
+                callback.load_state_dict(state)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1726]: same here

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:1671]: hm prefer separate PR if that sounds ok

File: tests/callbacks/test_callbacks.py [hunk 6]
```diff
@@ -164,3 +203,49 @@ def test_resume_incomplete_callbacks_list_warning(tmpdir):
...
+    def on_save_checkpoint(self, trainer, pl_module, checkpoint):
+        return {"old_state_precedence": 10}
+
+    def on_load_checkpoint(self, trainer, pl_module, callback_state):
+        assert callback_state == {"old_state_precedence": 10}
+        self.old_state_precedence = callback_state["old_state_precedence"]
+
+
+def test_resume_callback_state_all(tmpdir):
+    """Test on_save/load_checkpoint state precedence over state_dict/load_state_dict until v1.8 removal."""
```
Comment (awaelchli) [inline tests/callbacks/test_callbacks.py:232]: Great!

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 7]
```diff
@@ -386,6 +387,12 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
                 checkpoint[datamodule.__class__.__qualname__] = datamodule_state_dict
 
         # on_save_checkpoint hooks
+        if not weights_only:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:390]: I don't understand how this is backward compatible. If someone wants to return state until 1.8, they should be able to do so.

Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:390]: Yes returning state in `on_save_checkpoint` is supported (and overrides `state_dict`) until 1.8 (see `test_resume_callback_state_all` as well as `_call_callbacks_on_save_checkpoint` docstring)

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -1661,19 +1661,28 @@ def _on_train_batch_end(self, outputs: STEP_OUTPUT, batch, batch_idx, dataloader
...
+                callback_state_dicts[callback.state_key] = state_dict
+        return callback_state_dicts
+
+    def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """Called when saving a model checkpoint, calls every callback's `on_save_checkpoint` hook.
 
-        Calls every callback's `on_save_checkpoint` hook. We have a dedicated function for this rather than using
-        `_call_callback_hooks` because we have special logic for returning callback_states.
+        Will be removed in v1.8: If state is returned, we insert the callback state into
+        checkpoint["callbacks"][`Callback.state_key`]. It overrides `state_dict` if already present.
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1678]:
```suggestion
        ``checkpoint["callbacks"][Callback.state_key]``. It overrides ``state_dict`` if already present.
```

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -1661,19 +1661,28 @@ def _on_train_batch_end(self, outputs: STEP_OUTPUT, batch, batch_idx, dataloader
...
+            if state_dict:
+                callback_state_dicts[callback.state_key] = state_dict
+        return callback_state_dicts
+
+    def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """Called when saving a model checkpoint, calls every callback's `on_save_checkpoint` hook.
 
-        Calls every callback's `on_save_checkpoint` hook. We have a dedicated function for this rather than using
-        `_call_callback_hooks` because we have special logic for returning callback_states.
+        Will be removed in v1.8: If state is returned, we insert the callback state into
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1677]:
```suggestion
        Will be removed in v1.8: If state is returned, we insert the callback state into
```
lets format is deprecation `.. deprecated:: `

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (daniellepintz) [inline pytorch_lightning/trainer/trainer.py:1677]: will this whole method be removed in v1.8?

Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:1677]: not the whole method, just the part where we overwrite state:
https://github.com/PyTorchLightning/pytorch-lightning/blob/b5b951b05af8811f3e78636f17285f60a2a22e4f/pytorch_lightning/trainer/trainer.py#L1689-L1695

**Pull request merged**